### ISSUE TITLE
Native null assertions should be set to true by default in Dart 3

### ISSUE THREAD

**Issue by srujzs** [dart-lang/sdk#51799]
Further discussion and context in https://github.com/dart-lang/sdk/issues/50710.

The intended change in behavior: For null-safety, we've added the `--native-null-assertions` and `nativeNonNullAsserts` flags in the web compilers to make sure the web compilers were sound and to catch issues with the way we typed our `native` APIs. We made this flag default false before. This will now be default true in sound mode in Dart 3, and so users who are using sound null-safety already might come across runtime assertions with Dart 3. There are other changes in the above issue (like optimization flags), but those changes are not breaking.

The justification/rationale for making the change: We'd like to make the web libraries null-safe for users by *default*. Allowing unsoundness here may result in confusing errors further in the program, making it hard for users to debug.

The expected impact of this change: Flutter already enables these flags today, so it should see no change in behavior. google3 explicitly disables these flags, so it should see no change in behavior. The plan is to modify their defaults, but that's a separate change. Build web compilers users can see a difference here as it passes no default value, and therefore will use the compilers' defaults. Any other users not using these build systems can also possibly see a difference depending on their defaults. Users who might come across this will see runtime errors indicating that a particular API was expected to be non-nullable, but returned null. We expect these assertions to be rare in general.

Clear steps for mitigating the change: There's a doc for users (that DDC links to when users come across these assertions) to mitigate this: https://github.com/dart-lang/sdk/blob/main/sdk/lib/html/doc/NATIVE_NULL_ASSERTIONS.md. In order to disable this, users have to pass a flag either to the compilers or build web compilers.

**Comment by sigmundch** [comment]
/cc @itsjustkevin @mit-mit @vsmenon - to start the approval process.

I'm supportive of this breaking change. To clarify on the risk:
* It changes the default value of a flag, to align with our sound null safety changes in Dart 3.
* The mitigation is simple: pass a flag explicitly
* This breaks developers using build_web_compilers that run dart2js under the default optimization level (`-O2`)
* All other developers are not affected, including: flutter web, internal customers, and those using `-O3` or higher.

**Comment by itsjustkevin** [comment]
Tagging @Hixie and @grouma also.

**Comment by srujzs** [comment]
Friendly ping so we can get this in before the cutoff.

**Comment by mit-mit** [comment]
SGTM!

**Comment by grouma** [comment]
SGTM from a Google3 perspective.

**Comment by itsjustkevin** [comment]
I am marking this as approved, @Hixie if you see any issues here, please let us know.

**Comment by Hixie** [comment]
LGTM on the understanding that the only impact this has is on people directly using Dart's web compilers (not via Flutter).

**Comment by annagrin** [comment]
@nshahan @rakudrama checking the status - should this be closed now?

**Comment by srujzs** [comment]
Yes, this is landed and resolved. There is separate work to enable this internally, but I'll track that separately.
